app/src/main/java/org/thoughtcrime/securesms/net/ChunkedDataFetcher.java

```diff
@@ -0,0 +1,405 @@
+    if (contentLength <= 0) {
+      return fetchChunksWithUnknownTotalSize(url, callback);
+    }
+
+    CompositeRequestController compositeController = new CompositeRequestController();
+    fetchChunks(url, contentLength, Optional.absent(), compositeController, callback);
+    return compositeController;
+  }
+
+  private RequestController fetchChunksWithUnknownTotalSize(@NonNull String url, @NonNull Callback callback) {
+    CompositeRequestController compositeController = new CompositeRequestController();
+
+    long    chunkSize = new SecureRandom().nextInt(1024) + 1024;
+    Request request   = new Request.Builder()
+                                   .url(url)
+                                   .cacheControl(NO_CACHE)
+                                   .addHeader("Range", "bytes=0-" + (chunkSize - 1))
+                                   .addHeader("Accept-Encoding", "identity")
+                                   .build();
+
+    Call firstChunkCall = client.newCall(request);
+    compositeController.addController(new CallRequestController(firstChunkCall));
+
+    firstChunkCall.enqueue(new okhttp3.Callback() {
+      @Override
+      public void onFailure(@NonNull Call call, @NonNull IOException e) {
+        if (!compositeController.isCanceled()) {
+          callback.onFailure(e);
+          compositeController.cancel();
+        }
+      }
+
+      @Override
+      public void onResponse(@NonNull Call call, @NonNull Response response) {
+        String contentRange = response.header("Content-Range");
+
+        if (!response.isSuccessful()) {
+          Log.w(TAG, "Non-successful response code: " + response.code());
+          callback.onFailure(new IOException("Non-successful response code: " + response.code()));
+          compositeController.cancel();
+          if (response.body() != null) response.body().close();
+          return;
+        }
+
+        if (TextUtils.isEmpty(contentRange)) {
+          Log.w(TAG, "Missing Content-Range header.");
+          callback.onFailure(new IOException("Missing Content-Length header."));
+          compositeController.cancel();
+          if (response.body() != null) response.body().close();
+          return;
+        }
+
+        if (response.body() == null) {
+          Log.w(TAG, "Missing body.");
+          callback.onFailure(new IOException("Missing body on initial request."));
+          compositeController.cancel();
+          return;
+        }
+
+        Optional<Long> contentLength = parseLengthFromContentRange(contentRange);
+
+        if (!contentLength.isPresent()) {
+          Log.w(TAG, "Unable to parse length from Content-Range.");
+          callback.onFailure(new IOException("Unable to get parse length from Content-Range."));
+          compositeController.cancel();
+          return;
+        }
+
+        if (chunkSize >= contentLength.get()) {
+          try {
+            callback.onSuccess(response.body().byteStream());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *byteStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *byteStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).